File: Fun/Chess/ChessGame.java
package Chess;
import java.util.*;

public class ChessGame {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.println("Who is player one? [1] Human, [2] Robot, [3] Random");
        ChessPlayer p1 = new ChessPlayer(scanner.nextInt());

        System.out.println("Who is player two? [1] Human, [2] Robot, [3] Random");
        ChessPlayer p2 = new ChessPlayer(scanner.nextInt());

        int w1 = 0;
        int w2 = 0;
        int t = 0;

        for (int i = 0; i <1; i++) {
            System.out.println("Starting game                                                 " + (i + 1));
            char currentPlayer = 'w';
            boolean gameOver = false;
            ChessBoard board = new ChessBoard();
            Map<String, Integer> boardStates = new HashMap<String, Integer>();

            while (!gameOver) {
                board.printBoard();

                if (board.isInCheckmate(currentPlayer)) {
                    System.out.println((currentPlayer == 'w' ? "White" : "Black") + " is in checkmate. Game over!");
                    if (currentPlayer == 'w') {
                        w2++;
                    } else {
                        w1++;
                    }
                    gameOver = true;
                    continue;
                }
                if (boardStates.containsKey(board.toString()) && boardStates.get(board.toString()) >= 3) {
                    System.out.println("Draw by threefold repetition! Game over!");
                    t++;
                    gameOver = true;
                    continue;
                }

                if (board.isInStalemate(currentPlayer)) {
                    System.out.println("Stalemate! Game over!");
                    t++;
                    gameOver = true;
                    continue;
                }

                if (board.isFiftyMoveRule()) {
                    System.out.println("Draw by 50-move rule! Game over!");
                    t++;
                    gameOver = true;                 
                    continue;
                }

                if (board.isInsufficientMaterial()) {
                    System.out.println("Draw by insufficient material! Game over!");
                    t++;
                    gameOver = true;
                    continue;
                }

                if (board.isInCheck(currentPlayer)) {
                    System.out.println((currentPlayer == 'w' ? "White" : "Black") + " is in check!");
                }

                if (!gameOver) {
                    System.out.println((currentPlayer == 'w' ? "White" : "Black") + "'s move.");
                    String[] move;
                    if (currentPlayer == 'w') {
                        move = p1.getMove(board, 'w',boardStates,scanner);
                    } else {
                        move = p2.getMove(board, 'b',boardStates,scanner);
                    }

                    /*
                     try {
                         Thread.sleep(1000);
                     } catch (InterruptedException e) {
                         e.printStackTrace();
                     }
                    */

                    boolean moveResult = board.movePiece(move[0], move[1], currentPlayer, scanner);
                    if (moveResult) {
                        currentPlayer = (currentPlayer == 'w') ? 'b' : 'w';
                        if(boardStates.containsKey(board.toString())) {
                            boardStates.put(board.toString(), boardStates.get(board.toString()) + 1);
                        } else {
                            boardStates.put(board.toString(), 1);
                        }
                    } else {
                        System.out.println("Invalid move. Try again.");
                    }
                }
            }
        }

        System.out.println("Player one wins: " + w1);
        System.out.println("Player two wins: " + w2);
        System.out.println("Ties: " + t);

        scanner.close();
    }
}

File: site/src/Chess-js/ChessGame.js
class ChessGame {
    constructor() {
        this.w1 = 0;
        this.w2 = 0;
        this.t = 0;
    }

    startGame(p1Type, p2Type) {
        const p1 = new ChessPlayer(p1Type);
        const p2 = new ChessPlayer(p2Type);

        let currentPlayer = 'w';
        let gameOver = false;
        const board = new ChessBoard();
        const boardStates = new Map();

        return {
            board,
            currentPlayer,
            gameOver,
            boardStates,
            p1,
            p2
        };
    }

    playTurn(gameState) {
        const { board, boardStates, p1, p2, currentPlayer } = gameState;

        board.printBoard();

        // Check game end conditions
        if (board.isInCheckmate(currentPlayer)) {
            console.log((currentPlayer === 'w' ? "White" : "Black") + " is in checkmate. Game over!");
            console.log("The game was " + board.fullmoveNumber + " moves.");
            if (currentPlayer === 'w') {
                this.w2++;
            } else {
                this.w1++;
            }
            return { gameOver: true };
        }

        if (boardStates.has(board.toString())) {
            if (boardStates.get(board.toString()) >= 3) {
                console.log("Draw by threefold repetition! Game over!");
                this.t++;
                return { gameOver: true };
            }
        }

        if (board.isInStalemate(currentPlayer)) {
            console.log("Stalemate! Game over!");
            this.t++;
            return { gameOver: true };
        }

        if (board.isFiftyMoveRule()) {
            console.log("Draw by 50-move rule! Game over!");
            this.t++;
            return { gameOver: true };
        }

        if (board.isInsufficientMaterial()) {
            console.log("Draw by insufficient material! Game over!");
            this.t++;
            return { gameOver: true };
        }

        if (board.isInCheck(currentPlayer)) {
            console.log((currentPlayer === 'w' ? "White" : "Black") + " is in check!");
        }

        // Get player move
        const playerObj = currentPlayer === 'w' ? p1 : p2;
        const move = playerObj.getMove(board, currentPlayer, boardStates);

        // Execute move
        const moveResult = board.movePiece(move[0], move[1], currentPlayer, playerObj.getType());
        if (moveResult) {
            const nextPlayer = (currentPlayer === 'w') ? 'b' : 'w';
            if (boardStates.has(board.toString())) {
                boardStates.set(board.toString(), boardStates.get(board.toString()) + 1);
            } else {
                boardStates.set(board.toString(), 1);
            }
            return { gameOver: false, currentPlayer: nextPlayer };
        } else {
            console.log("Invalid move. Try again.");
            return { gameOver: false, currentPlayer };
        }
    }

    getResults() {
        return {
            player1Wins: this.w1,
            player2Wins: this.w2,
            ties: this.t
        };
    }
}